plotter/general_test.go

```diff
@@ -0,0 +1,128 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package plotter
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/gonum/plot"
+	"github.com/gonum/plot/vg"
+)
+
+var generateTestData = flag.Bool("regen", false, "Uses the current state to regenerate the test data.")
+
+// checkPlot checks a generated plot against a previously created reference.
+// If generateTestData = true, it regenerates the reference first.
+func checkPlot(dir, name, ext string, p *plot.Plot, width, height vg.Length,
+	errf errFunc, lf logFunc) {
+
+	filename := filepath.Join(dir, fmt.Sprintf("%s.%s", name, ext))
+
+	c, err := p.WriterTo(width, height, ext)
+	errf(err)
+
+	var buf bytes.Buffer
+	_, err = c.WriteTo(&buf)
+	errf(err)
+
+	// Recreate Golden images.
+	if *generateTestData {
+		errf(p.Save(width, height, filename))
+	}
+
+	f, err := os.Open(filename)
+	errf(err)
+
+	want, err := ioutil.ReadAll(f)
+	errf(err)
+	f.Close()
+	if !bytes.Equal(buf.Bytes(), want) {
+		errf(fmt.Errorf("image mismatch for %s\n", filename))
+		return
+	}
+	lf("Image can be seen at https://github.com/gonum/plot/tree/master/plotter/%s/%s.%s.\n"+
+		"Normally, you would use plot.Save().\n", dir, name, ext)
+}
+
+// errFunc is an interface for functions that deal with errors.
+type errFunc func(err error)
+
+// handleEx prints err if err!=nil
+var handleEx = func(err error) {
```

Comment:
A log.Fatalf would do the same thing and be more explanatory to the reader - which is the point of an example.

>It needs to be a var so it can be passed to checkPlot() but we create separate checkPlots for tests and examples then it wouldn't need to be. Same goes for below.

This is not true. A func can be used as a parameter to a function without being a declared variable. The issue here is that fmt.Printf returns (int, error), however that is a separate issue from the mode of declaration of the function.

http://play.golang.org/p/kTO9j7QEoR